### PR TITLE
Upgrade to PHP 7.0

### PR DESCRIPTION
**Pull request by jack-worman**
Closed other pull request because I could not fix the test failures, so I re-started and did the changes incrementally this time.

Overview:
* Added `declare(strict_types=1)`
* Added parameter types based on `@param` annotations
* Added return types based on `@return` annotations
* Increased PHP version to 7.1
  * Allowed me to add `void` returns and remove special PHPUnit setup
* Increased Psalm coverage to every file and put error level to 1
  * Baselined all existing errors
* Added php-cs-fixer migration rules

### CODE REVIEW

**Comment by terrafrost** [comment]
Thanks for this!

And sorry for the delay - I've been vacationing in California (Yosemite, Death Valley, etc). I'm back now so I'll take a look at this within the next few days! (it's a big commit so it'll prob take more than one day).

Thanks again!!

File: phpseclib/File/ASN1.php [hunk 1]
```diff
@@ -849,12 +839,10 @@ public static function encodeDER($source, $mapping, $special = [])
...
      *
      * @param Element|string|array|null $source
-     * @param array $mapping
-     * @param int $idx
-     * @param array $special
+     * @param string|int|null $idx
      * @return string
      */
-    private static function encode_der($source, $mapping, $idx = null, $special = [])
+    private static function encode_der($source, array $mapping, $idx = null, array $special = [])
```
Comment (terrafrost) [inline phpseclib/File/ASN1.php:845]: I feel like `int $idx = null` should work here?

File: phpseclib/Common/Functions/Strings.php [hunk 2]
```diff
@@ -295,11 +277,8 @@ public static function bits2bin($x)
 
     /**
      * Convert bits to binary data
-     *
-     * @param string $x
-     * @return string
      */
-    public static function bin2bits($x, $trim = true)
+    public static function bin2bits(string $x, $trim = true): string
```
Comment (terrafrost) [inline phpseclib/Common/Functions/Strings.php:281]: `$trim = true` prob ought to be changed to `bool $trim = true`

File: phpseclib/Crypt/EC/BaseCurves/Binary.php [hunk 3]
```diff
@@ -84,7 +86,7 @@ class Binary extends Base
     /**
      * Sets the modulo
      */
-    public function setModulo(...$modulo)
+    public function setModulo(...$modulo): void
```
Comment (terrafrost) [inline phpseclib/Crypt/EC/BaseCurves/Binary.php:89]: Should probably do `int ...$modulo` for this one

File: phpseclib/Crypt/EC/BaseCurves/Prime.php [hunk 4]
```diff
@@ -645,7 +642,7 @@ public function multiplyAddPoints(array $points, array $scalars)
      *
      * @return int[]
      */
-    private function getNAFPoints($point, $wnd)
+    private function getNAFPoints($point, $wnd): array
```
Comment (terrafrost) [inline phpseclib/Crypt/EC/BaseCurves/Prime.php:645]: That can be `array $point, int $wnd`

File: phpseclib/Crypt/RC2.php [hunk 5]
```diff
@@ -339,7 +333,7 @@ public function setKey($key, $t1 = false)
         }
 
         if ($t1 < 1 || $t1 > 1024) {
-            throw new \LengthException('Key size of ' . $length . ' bits is not supported by this algorithm. Only keys between 1 and 1024 bits, inclusive, are supported');
+            //throw new \LengthException('Key size of ' . $t1 . ' bits is not supported by this algorithm. Only keys between 1 and 1024 bits, inclusive, are supported');
```
Comment (terrafrost) [inline phpseclib/Crypt/RC2.php:336]: Why was this changed?

**Review by terrafrost — COMMENTED** [review]
I'm still working on this.

.github/workflows/ci.yml is updated so that the changes are only united tested in PHP 8.1 but in composer.json the minimum PHP version is bumped to 7.1?

I don't at all like all the hard-coded stuff in build/psalm_baseline.xml but as the minimum required version of PHP gradually increases a lot of these should be able to be removed.

Like PHP 8.0 introduced [union types](https://www.php.net/manual/en/language.types.declarations.php#language.types.declarations.composite.union), the `mixed` keyword, etc.

PHP 7.2 introduced `object` as a keyword (which might be best for Strings::is_stringable)

For some things, like PublicKeyLoader::load and all the associated methods (eg. Crypt/*/Formats/Keys/*.php), I feel like changing it to `string $password = null` from `$password = false` would be better. This could also be extended to the savePrivateKey functions as well, Hash::setKey, Hash::setNonce,

**Comment by terrafrost** [comment]
So I'm trying to backport the 5.6 compatible changes back to the 3.0 branch. Here's what I got so far:

https://github.com/terrafrost/phpseclib/tree/3.0-type-hinting-backport

Changing the function definition of ASN1::asn1map from `public static function asn1map($decoded, $mapping, $special = [])` to `public static function asn1map(array $decoded, $mapping, $special = [])` introduced some new issues. Like if ASN::decodeBER fails a false would be passed to ASN1::asn1map instead of an array. I didn't know that and fixed that up. There are still failing unit tests in my backports but I'll deal with them all in good time.

I'm not sure why those unit tests are passing in this PR but I'm still reviewing this PR so the answer may become apparent in due time.

I appreciate your patience!

File: phpseclib/Common/Functions/Strings.php [hunk 6]
```diff
@@ -295,11 +277,8 @@ public static function bits2bin($x)
 
     /**
      * Convert bits to binary data
-     *
-     * @param string $x
-     * @return string
      */
-    public static function bin2bits($x, $trim = true)
```
[the remark this answers is not in the document]
Comment (jack-worman) [reply phpseclib/Common/Functions/Strings.php:302]: Done

File: phpseclib/Crypt/RC2.php [hunk 7]
```diff
@@ -325,12 +321,10 @@ public function getKeyLength()
...
      * it is empty.
      *
-     * @see \phpseclib3\Crypt\Common\SymmetricKey::setKey()
-     * @param string $key
-     * @param int|boolean $t1 optional Effective key length in bits.
      * @throws \LengthException if the key length isn't supported
+     *@see \phpseclib3\Crypt\Common\SymmetricKey::setKey()
      */
-    public function setKey($key, $t1 = false)
+    public function setKey(string $key, $t1 = false): void
```
[the remark this answers is not in the document]
Comment (jack-worman) [reply phpseclib/Crypt/RC2.php:327]: Did `int $t1 = null`

File: phpseclib/Crypt/RSA.php [hunk 8]
```diff
@@ -499,10 +493,9 @@ protected function __construct()
      * See {@link http://tools.ietf.org/html/rfc3447#section-4.1 RFC3447#section-4.1}.
      *
      * @param bool|\phpseclib3\Math\BigInteger $x
-     * @param int $xLen
      * @return bool|string
      */
-    protected function i2osp($x, $xLen)
+    protected function i2osp($x, int $xLen)
```
[the remark this answers is not in the document]
Comment (jack-worman) [reply phpseclib/Crypt/RSA.php:498]: Done

File: phpseclib/Crypt/Twofish.php [hunk 9]
```diff
@@ -525,12 +523,10 @@ protected function setupKey()
...
      * _mdsrem function using by the twofish cipher algorithm
-     *
-     * @param string $A
-     * @param string $B
-     * @return array
+     * @param string|int $A
+     * @param string|int $B
      */
-    private function mdsrem($A, $B)
+    private function mdsrem($A, $B): array
```
[the remark this answers is not in the document]
Comment (jack-worman) [reply phpseclib/Crypt/Twofish.php:529]: Done

File: phpseclib/File/ASN1.php (hunk 1 above)
Comment (jack-worman) [reply phpseclib/File/ASN1.php:845]: Tests fail because a string is passed to `$idx`.

File: phpseclib/Common/Functions/Strings.php (hunk 2 above)
Comment (jack-worman) [reply phpseclib/Common/Functions/Strings.php:281]: Done

File: phpseclib/Crypt/EC/BaseCurves/Binary.php (hunk 3 above)
Comment (jack-worman) [reply phpseclib/Crypt/EC/BaseCurves/Binary.php:89]: Done

File: phpseclib/Crypt/EC/BaseCurves/Prime.php (hunk 4 above)
Comment (jack-worman) [reply phpseclib/Crypt/EC/BaseCurves/Prime.php:645]: Done

File: phpseclib/Crypt/RC2.php (hunk 5 above)
Comment (jack-worman) [reply phpseclib/Crypt/RC2.php:336]: `$length` is an undefined variable.
As for why it was commented out, I'm not sure.

Uncommented.

**Comment by jack-worman** [comment]
> .github/workflows/ci.yml is updated so that the changes are only united tested in PHP 8.1 but in composer.json the minimum PHP version is bumped to 7.1?

My thinking was: why bother with testing on other versions when they will be going away soon.

> I don't at all like all the hard-coded stuff in build/psalm_baseline.xml but as the minimum required version of PHP gradually increases a lot of these should be able to be removed.

What hard coded stuff are you talking about? If you mean all the added errors, you can justify it by knowing these were all pre-existing and psalm was just (implicitly) ignoring them before.

> For some things, like PublicKeyLoader::load and all the associated methods (eg. Crypt//Formats/Keys/.php), I feel like changing it to string $password = null from $password = false would be better. This could also be extended to the savePrivateKey functions as well, Hash::setKey, Hash::setNonce,

In a different issue please. Let's not add scope to this giant PR 😛

File: phpseclib/Math/BinaryField/Integer.php [hunk 10]
```diff
@@ -447,40 +429,32 @@ public function negate()
 
     /**
      * Returns the modulo
-     *
-     * @return string
      */
-    public static function getModulo($instanceID)
+    public static function getModulo($instanceID): string
```
Comment (terrafrost) [inline phpseclib/Math/BinaryField/Integer.php:433]: Should be able to replace `$instanceID` with `int $instanceID`

File: phpseclib/Math/BinaryField/Integer.php [hunk 11]
```diff
@@ -160,7 +155,7 @@ private static function deg($x)
      * @return string[]
      * @link https://en.wikipedia.org/wiki/Polynomial_greatest_common_divisor#Euclidean_division
      */
-    private static function polynomialDivide($x, $y)
+    private static function polynomialDivide($x, $y): array
```
Comment (terrafrost) [inline phpseclib/Math/BinaryField/Integer.php:158]: Should be able to do `string $x, string $y`

File: phpseclib/Math/BinaryField/Integer.php [hunk 12]
```diff
@@ -187,10 +182,9 @@ private static function polynomialDivide($x, $y)
     /**
      * Perform polynomial multiplation in the traditional way
      *
-     * @return string
      * @link https://en.wikipedia.org/wiki/Finite_field_arithmetic#Multiplication
      */
-    private static function regularPolynomialMultiply($x, $y)
+    private static function regularPolynomialMultiply($x, $y): string
```
Comment (terrafrost) [inline phpseclib/Math/BinaryField/Integer.php:187]: Should be able to do `string $x, string $y`

File: phpseclib/Math/BinaryField/Integer.php [hunk 13]
```diff
@@ -226,10 +220,9 @@ private static function regularPolynomialMultiply($x, $y)
      *
      * Uses karatsuba multiplication to reduce x-bit multiplications to a series of 32-bit multiplications
      *
-     * @return string
      * @link https://en.wikipedia.org/wiki/Karatsuba_algorithm
      */
-    private static function polynomialMultiply($x, $y)
+    private static function polynomialMultiply($x, $y): string
```
Comment (terrafrost) [inline phpseclib/Math/BinaryField/Integer.php:225]: Should be able to do `string $x, string $y`

File: phpseclib/Math/PrimeField/Integer.php [hunk 14]
```diff
@@ -102,19 +94,16 @@ public static function setRecurringModuloFunction($instanceID, callable $functio
     /**
      * Delete the modulo for a given instance
      */
-    public static function cleanupCache($instanceID)
+    public static function cleanupCache($instanceID): void
```
Comment (terrafrost) [inline phpseclib/Math/PrimeField/Integer.php:97]: Should be able to do `int $instanceID`

File: phpseclib/Net/SFTP.php [hunk 15]
```diff
@@ -829,11 +805,9 @@ private function nlist_helper($dir, $recursive, $relativeDir)
     /**
      * Returns a detailed list of files in the given directory
      *
-     * @param string $dir
-     * @param bool $recursive
      * @return array|false
      */
-    public function rawlist($dir = '.', $recursive = false)
+    public function rawlist(string $dir = '.', bool $recursive = false)
```
Comment (terrafrost) [inline phpseclib/Net/SFTP.php:810]: Prob ought to update these to return null on failure vs false. Kinda like what I proposed with `PublicKeyLoader::load`. Can be done with a later PR however.

File: phpseclib/Net/SFTP.php [hunk 16]
```diff
@@ -980,11 +952,9 @@ private function readlist($dir, $raw = true)
      *
      * Intended for use with uasort()
      *
-     * @param array $a
-     * @param array $b
      * @return int
      */
-    private function comparator($a, $b)
+    private function comparator(array $a, array $b)
```
Comment (terrafrost) [inline phpseclib/Net/SFTP.php:957]: In reviewing the code it doesn't look like this returns a mixed type - looks like it just returns an int so we should prob add that as a return type hint.

File: phpseclib/Net/SFTP.php [hunk 17]
```diff
@@ -1294,12 +1253,10 @@ public function lstat($filename)
      * Determines information without calling \phpseclib3\Net\SFTP::realpath().
      * The second parameter can be either PacketType::STAT or PacketType::LSTAT.
      *
-     * @param string $filename
-     * @param int $type
-     * @throws \UnexpectedValueException on receipt of unexpected packets
      * @return array|false
+     *@throws \UnexpectedValueException on receipt of unexpected packets
```
Comment (terrafrost) [inline phpseclib/Net/SFTP.php:1257]: I feel like there should be a space between `*` and `@throws`?

File: phpseclib/Net/SFTP.php [hunk 18]
```diff
@@ -1337,13 +1290,9 @@ public function truncate($filename, $new_size)
      *
      * If the file does not exist, it will be created.
      *
-     * @param string $filename
-     * @param int $time
-     * @param int $atime
-     * @throws \UnexpectedValueException on receipt of unexpected packets
-     * @return bool
+     *@throws \UnexpectedValueException on receipt of unexpected packets
```
Comment (terrafrost) [inline phpseclib/Net/SFTP.php:1293]: Seems like a space should be added here as well?

File: phpseclib/Net/SFTP.php [hunk 19]
```diff
@@ -1685,12 +1613,9 @@ public function readlink($link)
      *
      * symlink() creates a symbolic link to the existing target with the specified name link.
      *
-     * @param string $target
-     * @param string $link
-     * @throws \UnexpectedValueException on receipt of unexpected packets
-     * @return bool
+     *@throws \UnexpectedValueException on receipt of unexpected packets
```
Comment (terrafrost) [inline phpseclib/Net/SFTP.php:1616]: Space here as well?

**Review by terrafrost — COMMENTED** [review]
Still working on this!

File: phpseclib/Net/SFTP.php (hunk 19 above)
Comment (jack-worman) [reply phpseclib/Net/SFTP.php]: Fixed all

File: phpseclib/Net/SFTP.php (hunk 18 above)
Comment (jack-worman) [reply phpseclib/Net/SFTP.php]: Fixed all

File: phpseclib/Net/SFTP.php (hunk 17 above)
Comment (jack-worman) [reply phpseclib/Net/SFTP.php]: Fixed all

File: phpseclib/Net/SFTP.php (hunk 16 above)
Comment (jack-worman) [reply phpseclib/Net/SFTP.php]: Done

File: phpseclib/Math/BinaryField/Integer.php (hunk 13 above)
Comment (jack-worman) [reply phpseclib/Math/BinaryField/Integer.php]: Done

File: phpseclib/Math/BinaryField/Integer.php (hunk 12 above)
Comment (jack-worman) [reply phpseclib/Math/BinaryField/Integer.php]: Done

File: phpseclib/Math/BinaryField/Integer.php (hunk 11 above)
Comment (jack-worman) [reply phpseclib/Math/BinaryField/Integer.php]: Done

File: phpseclib/Math/BinaryField/Integer.php (hunk 10 above)
Comment (jack-worman) [reply phpseclib/Math/BinaryField/Integer.php]: Done

**Comment by jack-worman** [comment]
Addressed 2nd round of comments

**Comment by jack-worman** [comment]
Rebased onto master

File: phpseclib/File/X509.php [hunk 20]
```diff
@@ -429,12 +430,8 @@ public function __construct()
      * Load X.509 certificate
      *
      * Returns an associative array describing the X.509 cert or a false if the cert failed to load
-     *
-     * @param string $cert
-     * @param int $mode
-     * @return mixed
      */
-    public function loadX509($cert, $mode = self::FORMAT_AUTO_DETECT)
+    public function loadX509(string $cert, int $mode = self::FORMAT_AUTO_DETECT)
```
Comment (terrafrost) [inline phpseclib/File/X509.php:434]: This needs to be able to accept arrays. The code right below it does an `is_array($cert)` check and then does some specific things if it is an array. https://github.com/phpseclib/phpseclib/commit/3f9aa1ad6a6a1ca4ce98fb1f712f4c8e0a471bf6 is the change that added that.

I suppose we could drop support for arrays but idk that I see an overwhelming need to

File: phpseclib/System/SSH/Common/Traits/ReadBytes.php [hunk 21]
```diff
@@ -23,10 +25,9 @@ trait ReadBytes
     /**
      * Read data
      *
-     * @param int $length
      * @throws \RuntimeException on connection errors
      */
-    public function readBytes($length)
+    public function readBytes(int $length)
```
Comment (terrafrost) [inline phpseclib/System/SSH/Common/Traits/ReadBytes.php:30]: We should prob add a return type hint of string to this? If `fread` returns anything other than a string an exception prob ought to be thrown

**Comment by terrafrost** [comment]
My code review of this is complete. I merged the backport in to the 3.0 branch and merged that into master. Looks like that's introduced some merge conflicts. Fix those up and look into my comments and then I'll merge this.

Thank you for your patience!
